### PR TITLE
chore: upgrade to yarn 3

### LINKED ISSUE #6150: Repo: Can't develop because of yarn1
### Suggestion

There are some issues with yarn1 that will never be fixed, such as not being able to run any commands with yarn1 when there are unicode characters in the windows username.
This pretty much means that these users (like me) can't build, debug and develop at all.

Migrating to yarn3 seems like a good approach, it is actively maintained.

### PR DESCRIPTION
**Pull request by JamesHenry**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6150 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Migrates the codebase from yarn v1 to yarn v3, using the `node_modules` linker - **not** PnP - to ensure maximum consistency before and after the migration.

I confirmed that there were no issues with lerna publishing by publishing a canary version to verdaccio locally (this will not negatively impact the canary release post-merge):

image

### CODE REVIEW

File: packages/ast-spec/project.json
Comment (JamesHenry) [inline packages/ast-spec/project.json:13]: Unrelated to yarn 3

File: packages/shared-fixtures/project.json
Comment (JamesHenry) [inline packages/shared-fixtures/project.json:2]: Unrelated to yarn 3

File: packages/website-eslint/project.json
Comment (JamesHenry) [inline packages/website-eslint/project.json:2]: Unrelated to yarn 3

**Comment by JamesHenry** [comment]
@bradzacher @JoshuaKGoldberg please have a play with this, it all looks good to me locally and CI is passing.

Once you are happy and approve the PR please **do not merge**, so that I can coordinate the update of the automation jobs

File: packages/ast-spec/project.json
Comment (JoshuaKGoldberg) [reply packages/ast-spec/project.json:13]: Random thought: one could write Nx-specific lint rules for `project.json` files. There could be a lint rule that suggests replacing hardcoded paths like `"packages/ast-spec` with `"${projectRoot}"`.

File: packages/eslint-plugin-internal/package.json
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin-internal/package.json:24]: Kind of nice to have these more explicit now. +1

File: tools/postinstall.ts
Comment (JoshuaKGoldberg) [inline tools/postinstall.ts:7]: I ran `yarn dedupe` and it generated `21 insertions(+), 564 deletions(-)`. Maybe we can add `yarn dedupe` here too?

**Review by JoshuaKGoldberg — APPROVED** [review]
Looks great, thanks! 🔥

File: .gitignore
Comment (JoshuaKGoldberg) [inline .gitignore:90]: Can I suggest sending these in as a small PR now, before this one gets merged? That way we don't keep having 1k+ files left over when switching between this branch and others.

File: packages/ast-spec/project.json
Comment (JamesHenry) [reply packages/ast-spec/project.json:13]: So this is actually the result of running `nx repair`, as special form of migration that can be run any time to correct any legacy configuration issues or things which are _definitely_ better as far as the tooling can figure out. So I think it's more a case of figuring out a good way to periodically run `nx repair`. I think my existing automated workflow for version migrations makes the most sense for that and I'll add it

File: .gitignore
Comment (JamesHenry) [reply .gitignore:90]: Done

File: tools/postinstall.ts
Comment (JamesHenry) [reply tools/postinstall.ts:7]: We can't do it in the postinstall I don't think otherwise we end up in a loop when there are changes resulting from the dedupe.

I have pushed a dedicated commit with an alternative approach to ensuring that the lock file is maximally deduped over time: https://github.com/typescript-eslint/typescript-eslint/pull/6162/commits/18657c0f61619f1ef408ccd4c1ae095d55a239e1

File: packages/ast-spec/project.json
Comment (JamesHenry) [reply packages/ast-spec/project.json:13]: Added https://github.com/typescript-eslint/typescript-eslint/pull/6162/commits/b549f4e233cc76b21e34368765441b3e41c68337

File: tools/postinstall.ts
Comment (bradzacher) [inline tools/postinstall.ts:9]: side note that we could convert this to use zx and it would be a little better than maintaining our own exec version

https://github.com/google/zx

File: tools/postinstall.ts
Comment (bradzacher) [reply tools/postinstall.ts:7]: we can probably setup an action that runs weekly or something to generate a PR?
actually can we add it to the renovate config so it does it automatically? most of our dep upgrades come via dependabot anyways

**Review by bradzacher — DISMISSED** [review]
25a

once you fix the CI I'm happy with this.
now i'll need to get used to the new console output from yarn sadge

File: tools/postinstall.ts
Comment (JamesHenry) [reply tools/postinstall.ts:7]: > we can probably setup an action that runs weekly or something to generate a PR?

Hmm yeah that's probably going to be a lot less disruptive to contributors. Will be annoying to put together a PR and have it fail if you didn't prune recently... I'll revert that change and we can do a prune workflow in a follow up

> most of our dep upgrades come via dependabot anyways

@bradzacher isn't it the opposite? We switched from dependabot to renovate, right?

Comment (bradzacher) [reply tools/postinstall.ts:7]: yes, that's what I meant! I totally said both on purpose to cover my bases

Comment (JamesHenry) [reply tools/postinstall.ts:7]: (PS I've removed the call to `yarn dedupe` from ci.yml in this PR now so that we don't cause issues for users unnecessarily per the above comment)

**Review by bradzacher — APPROVED** [review]

File: tools/postinstall.ts
Comment (bradzacher) [reply tools/postinstall.ts:7]: See how forward thinking i am? I think so far ahead I forget how far ahead I think.

bigbraintime

**Review by bradzacher — APPROVED** [review]

**Comment by liuxingbaoyu** [comment]
Thanks for all this!

File: jest.config.base.js
Comment (fisker) [inline jest.config.base.js:24]: `<rootDir>/node_modules/console-fail-test/setup.js` should work.

Comment (bradzacher) [reply jest.config.base.js:24]: I think the issue with using `<rootDir>` is that it's dependent where the `jest.config.js` is.
If you `cd` into one of the packages, each package folder has its own `jest.config.js`, so that means that in a package `<rootDir>` will mean `<root>/packages/foo`, not just `<root>` - i.e. the path will not exist.

By using `_dirname` it means this path is always relative to this root config file.

File: package.json
Comment (JamesHenry) [inline package.json:63]: This was outdated, the package is `nx-cloud` now which we already have specified

File: package.json
Comment (JamesHenry) [inline package.json:1]: Yarn automatically reordered these deps (apart from the `@nrwl/nx-cloud` one, see note)

File: packages/website-eslint/package.json
Comment (JamesHenry) [inline packages/website-eslint/package.json:34]: These packages had been duplicated with the `dependencies` and were not being updated, causing issues. Just needed to remove them

File: packages/type-utils/package.json
Comment (JamesHenry) [inline packages/type-utils/package.json:55]: I had to downlevel this to match the rest of the workspace, otherwise `patch-package` hard errors on it

File: packages/utils/package.json
Comment (JamesHenry) [inline packages/utils/package.json:82]: Having to match all these `jest` versions is pretty cumbersome, but required to make the naked `jest` calls in npm scripts work with modern yarn.

The way they recommend handling this is different and involves sharing scripts between workspaces, or using `run -T` in front of the script contents.

Somewhat related - we could avoid the need for this entirely if we always let Nx run the tasks through its own executors and then just switch the script contents to be `nx test` like we did here: https://github.com/typescript-eslint/typescript-eslint/blob/ed8c6247c0454f84b126f5fbe16fe1c40f6605f3/packages/scope-manager/package.json#L43 - it would resolve `jest` for us, and we wouldn't need to bend over backwards to make yarn happy

File: packages/eslint-plugin/package.json
Comment (JamesHenry) [inline packages/eslint-plugin/package.json:79]: Changed to match other references

File: .github/workflows/nx-migrate.yml
Comment (JoshuaKGoldberg) [inline .github/workflows/nx-migrate.yml:76]: wat

https://github.com/nrwl/nx/issues/18385

Comment (JoshuaKGoldberg) [reply .github/workflows/nx-migrate.yml:76]: (not a blocker, I just am curious what this actually does)

File: netlify.toml
Comment (JoshuaKGoldberg) [inline netlify.toml:8]: That awkward moment when the upstream repo is archived without resolving the issue... 🤷

File: packages/utils/package.json
Comment (JoshuaKGoldberg) [reply packages/utils/package.json:82]: That sounds good to me! IIRC you'd mentioned this as a potential followup a while back. I'm definitely in favor.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
This looks great! I read through and think I understand the strategies and upstream-blocked-todo-workarounds. Also I tried it out locally and -with the caveat I commented in `package.json`- it seems to work great on my end. Nice! 👏 

Only requesting changes because I think we'll need to update the local development docs.

**Comment by JamesHenry** [comment]
@JoshuaKGoldberg Brought it back up to date and updated the local development guide. Hopefully we can merge now!

**Review by JoshuaKGoldberg — APPROVED** [review]
Tried it out locally again and it all works great! No need for me to manually run `.cjs` files from the repo. Thanks! Awesome stuff @JamesHenry 🚀

**Pull request merged**
